### DISCUSSION TITLE
superClass support for c++

### DISCUSSION THREAD

**Discussion by marekdaniluk** [antlr/antlr4#4434] [Q&A]
Hello,
I'm trying to create a cpp version of [YarnSpinner.Compiler/YarnSpinnerLexer.g4](https://github.com/YarnSpinnerTool/YarnSpinner), a branching dialog system designed in C Sharp. It uses antlr under the hood, so I tried it out in my C++ project.
I generated the source files using the `-Dlanguage=Cpp` option. YarnSpinner in the lexer script uses the `superClass` option which generates an incorrect `YarnSpinnerLexer.h` file. The problem is, that the YarnSpinnerLexer class inherits from another class not added by `#include`.
I am currently editing the generated header and adding a line with the appropriate include, but this solution seems inadequate. I wonder if there is a recommended way to solve similar problems for C++ runtime?

**Comment by kaby76** [comment]
You're going to have to add to the lexer .g4 file, just after the _options_ clause, `@header {#include "whateveryourfileis.h"}`.

Make sure to use a build process for generating the parser for the Cpp target automatically via the tool (CMake, Make, etc). You don't want to get in the habit of running the Antlr tool by hand. I noticed that the YarnSpinner.Compiler.csproj references the [Antlr4.Runtime.Standard](https://github.com/YarnSpinnerTool/YarnSpinner/blob/1fae2a59f8b910495354709d574635f2c7896729/YarnSpinner.Compiler/YarnSpinner.Compiler.csproj#L23) but does not include build support, e.g., [Antlr4BuildTasks](https://www.nuget.org/packages/Antlr4BuildTasks/). This means that someone manually runs the Antlr4 Tool by hand, which can lead to problems if not done correctly.

**Reply by phoenicyan** [reply]
Thank you for the @header! This should be masked as Answered.
